Add test.sh script for more consistent testing

techgique opened the pull request:
Add test.sh script for more consistent testing
  - Closes #535 

## Submitter Checklist

- [x] Verify all tests pass
  - Run tests with `docker-compose -f test-compose.yml -p onitest up test`
- [ ] Update `README.md` and `docs/` as necessary
  - [ ] If a `manage.py` command is added, deleted, or modified its help text must
    be valid and it should be documented in detail in
    `docs/advanced/admin-commands.md`
- Update `CHANGELOG.md`
  - [ ] Describe change(s) in appropriate section(s)
  - [ ] List self in Contributors section
- If a release PR:
  - [ ] In `CHANGELOG.md`, replace `[Unreleased]` with version and update compare link
  - [ ] Update `core/version.py` with new version
- [ ] Resolve merge conflicts
- [ ] @mention individual(s) you would like to review the PR
  - Reviews for releases must come from a reviewer at another institution

## Reviewer Checklist

- [ ] Verify all tests pass
- [ ] Review changes for behavior, bugs, and compliance with [Development
  Standards](https://github.com/open-oni/open-oni/tree/dev/CONTRIBUTING.md#development-standards)
  - Request the submitter make any changes rather than pushing changes yourself.
    Otherwise ask another reviewer to look at any changes you have made.
- [ ] If a release, follow the [Release
  Checklist](https://github.com/open-oni/open-oni/tree/dev/CONTRIBUTING.md#release-checklist)
<!-- Markdown renders in unwanted carriage return if this text is continued on
     the next line, so breaking character margin intentionally here -->

techgique committed:
Add test.sh script for more consistent testing

jechols commented:
I think for CI-style testing this general approach makes sense, but for local dev, this would get extremely burdensome.  When I do a "test loop" right now (where I leave the test containers up as I edit things), I can edit the code and then see if tests pass in a matter of about 30 seconds.  Removing every piece of persistence puts this closer to 10 minutes per test run.  For this reason I suggest any shortcuts we take include a separation of "CI" style testing from "dev test loop" approaches.

If we want a CI-style shortcut, this script is a good start, but there are a couple issues:

- The current test container mounts the code into the container and then copies it, rather than running it directly, in order to ensure things like themes aren't copied.  If we want to get ENV to be clean during tests, we could just add "ENV" to the list of excluded directories in the rsync in `docker/test-startup.sh`.  This would keep `ENV` for dev, but ensure that when the test container is started up fresh, it fully rebuilds `ENV`.
- Removing local images may or may not actually do anything.  Since docker is going to be caching pieces of the image build process, if you build a dev *and* a test image, you have two tags pointing to the same underlying image.  Telling docker to destroy the test image only untags it.  If we want to ensure a totally from-scratch image build, we need to do this differently, such as telling it to force-rebuild the image: `docker-compose -f test-compose.yml -p onitest build --no-cache`.  This is likely what we want for a true CI environment anyhow.

I'm of the opinion that we aren't going to make a one-size-fits-all test script, because of the significant difference between CI and quick local testing.  I'd like to see us focus more on documentation of local dev, including various test patterns, rather than trying to make a single path to do all tests.

techgique commented:
This is intended to make testing for new contributors easier, and for us to be more consistent in our local dev testing for core Open ONI because I'd seen almost half a dozen different commands pasted into Slack over the past week, and I want to come to a shared solution on that end. So all the CI stuff is kind of tangential to what I am aiming for here.

The point about removing local images makes sense. I had looked at the documentation and saw that removing all images is the other option, but I wasn't certain what we wanted and just wanted to get the draft PR up for discussion. Sounds like we will probably want the `all` option to completely test from scratch.

> I'd like to see us focus more on documentation of local dev, including various test patterns

This is what I want to avoid. It's a hassle to have to go read documentation to copy out commands to run tests in different situations. Even copying out the long command in our checklist on PRs feels unfriendly to me. I'm trying to handle the common uses I expect with the options on this script.

My thinking was to make one-off testing for non-core contributors who we are asking to run tests in the PR checklist be handled by the default run of this script. Wipe everything just in case there's anything remaining from the past, rebuild completely from scratch to be as thorough as possible, and wipe everything out again afterwards so we didn't just fill up a bunch of disk space on their system and leave things there. I want to practice good housekeeping by default.

I hadn't thought of this before (and haven't gone hunting into the code as a tangent while I'm writing out my thoughts), but can we react to the exit code from the tests such that if a test fails, we don't wipe everything out afterwards so the complete rebuild process isn't required again if a test failed? In that case we could output a message recommending the next run of the script uses the option to not clear the environment before running the tests again.

My intention was to handle the case of wanting a shorter test loop with the options one can pass to the script, particularly the `-p` option to persists the images, volumes, and containers. And the complimentary `-c` option is for running after short test loops are no longer needed and one is ready to clear the test environment and be done.

I now realize that I made an oversight in how I wrote the script because running the script as `test.sh -p` a second time will still clear the environment before running the tests. So I'll push a commit to fix that and switch to removing `all` images.

techgique commented:
Another possibility that crossed my mind is to not _completely_ wipe the Docker environment unless core contributors are testing a release PR and only clear containers and volumes for other use cases. I don't know if full image rebuilds are really necessary for most other test runs. Thoughts?

techgique committed:
Fix persist still clearing test env before tests

- If one passes both clear and persist options, nothing is done before
  exit. I hope we're okay not spelling that out to the user

techgique committed:
Switch clear_env to remove all images

jechols commented:
The `-p` flag fix is a good move for sure, but IMO that should be a default, and everything else should be optional.

What is a "clean" build?  For me, it often means `git clean -nxfd`, then some targeted `rm`s, and nothing more.  Sometimes it's just a rebuild, e.g., `docker-compose build web`.  Sometimes it really does mean a full refresh of the ONI image, but what you've got here won't do that.  As I mentioned before, even in the case that you want a completely "from scratch" build, there are issues with this script:

- Removing `ENV` shouldn't happen - that only affects development, not testing.  Testing never uses the local `ENV` dir.
- Removing images doesn't offer any guarantees of a clean environment.  Docker caches image layers aggressively.  `--rmi all` doesn't remove those cached elements.  The only way to ensure the ONI image is 100% up-to-date is to pull the latest ubuntu image and then build ONI with `--no-cache`.  I can elaborate further if necessary, but probably not on the ticket comments.

A from-scratch build should happen once per dev session **at most**, and if it happens, it should be at the beginning of the session, not when testing happens (which for some is after they've pushed up their code).  There's no hard and fast rule for when this should be done, though, so putting it into a specific context (testing) isn't a good idea.

If the goal is to ease contributions (which it sounds like is your goal here), let's make a `test.sh` that does the simplest, fastest testing possible by default; core devs can do further testing on a merge based on the changes made.  This still means documenting testing strategies, at least for the core devs.  And the nuances of when to do refreshes of images, data, etc. are still something serious contributors will need to know.

Making it easy to push up a quick fix is fine, but serious contributors need documentation that explains how and when to do in-depth refreshes, testing, dev, etc.

jechols commented:
> Another possibility that crossed my mind is to not _completely_ wipe the Docker environment unless core contributors are testing a release PR and only clear containers and volumes for other use cases. I don't know if full image rebuilds are really necessary for most other test runs. Thoughts?

I think this gets into another nuanced point here.  I think this makes sense, but any big change may need a very in-depth test run even before it gets into dev just because it's easier to fix things before they're affecting all devs.  I think it's just got to be a judgment call on the person doing the merge to decide what level of testing is needed.  With front-end-heavy changes, our automated tests won't do a thing, and refreshing images doesn't really matter; somebody just has to sit down and poke the site and make sure things seem sane.  There just isn't a single one-size-fits-all approach.

jduss4 commented:
If this script is targeted towards new contributors, having a test that automatically scrubs a lot of stuff and requires minutes to run is probably not so friendly, especially if most of the changes are to basic front-end type of stuff that shouldn't require rebuilding the environment.  As @jechols mentions, though, a lot of those front-end changes probably aren't really caught by the tests at the moment.  :/

I do like the idea of a more friendly / standardized test script, though.  I have appreciated having the prompt in the PR checklist, because that's easy to copy over and run.  I think I would prefer if this script started with a basic "just run tests as is" baseline and then flags would add complexity / destruction to it.  Then our PR checklist would just need to point out which flags might be appropriate ("run `test.sh`, adding -x, -y if code modifies Z")
